### ISSUE TITLE
"DNSSEC Validated" icon near DNSSEC incorrect domain

### ISSUE THREAD

**Issue by DavidOsipov** [AdguardTeam/AdGuardHome#3017]
### System Details:
* **Version of AdGuard Home server:**
  * Adguard Home v0.106.0-b.4
* **How did you install AdGuard Home:**
  * Snap package
* **How did you setup DNS configuration:**
  * Ubuntu VPS
* **CPU architecture:**
  * x86_64
* **Operating system and version:**
  * Ubuntu 20.04.2 LTS
* **Unbound is present:**
  * Unbound  1.13.0 (with DNSSEC features and hardenings on)
 
### Describe the bug
_I use Unbound + Adguard Home  
I turned on "Enable DNSSEC" feature in Adguard home
I use [www.servfail.sidnlabs.nl](https://www.servfail.sidnlabs.nl) and [www.dnssec-failed.org](https://www.dnssec-failed.org) domains to check if DNSSEC checks are really working_

**When DNSSEC checks are <ins>turned on</ins> in Unbound and Adguard Home:**

_Expected behavior:_
 No "DNSSEC Validated" icon near [www.servfail.sidnlabs.nl](https://www.servfail.sidnlabs.nl) and [www.dnssec-failed.org](https://www.dnssec-failed.org) domains in Adguard Home, because Unbound has restricted us from entering these domains with SERVFAIL

_Actual behavior:_
![DNSSEC3](https://user-images.githubusercontent.com/38164582/116271173-440f6480-a76f-11eb-85e1-6d5f6ce5d404.PNG)
![DNSSEC](https://user-images.githubusercontent.com/38164582/116271203-48d41880-a76f-11eb-985b-5f11983511ef.PNG)

**When DNSSEC checks are <ins>turned off</ins> in Unbound and <ins>turned on</ins> in Adguard Home:**

_Expected behavior:_
Domains [www.servfail.sidnlabs.nl](https://www.servfail.sidnlabs.nl) and [www.dnssec-failed.org](https://www.dnssec-failed.org) are inaccessible (SERVFAIL) and no "DNSSEC Validated" icon near them

_Actual behavior:_
![DNSSEC 1](https://user-images.githubusercontent.com/38164582/116272229-458d5c80-a770-11eb-831c-06251cc4a2ce.PNG)

**Comment by ameshkov** [comment]
Yeah, the icon is indeed misleading and we'd better improve it. And not just it, DNSSEC check should be implemented fully and not as it is now: AGH does not perform the validation by itself, it just asks the upstream resolver to do it. The icon indicates the fact that it asked to validate but does not indicate the fact that the validation failed.

**Comment by ainar-g** [comment]
Additional suggestions from #4258:

 >  Please add a tool-tip for the DNSSEC Icon in the Query Log when DNSSEC is not used (grey icon).
 >
 >   *  Has the it been resolved from local cache?
 >   *  Has it been affected by "optimistic caching"?
 >   *  Has it incorrectly not used DNSSEC when it should have?

In any case, those indicators could probably be more informative.

**Comment by DavidOsipov** [comment]
I have found out, that the optimistic cache is serving outdated DNSSEC signed responses, which are considered bogus by other DNSSEC validating software like Mozilla Thunderbird DKIM Verifier add-on. Not really a big deal, but can cause troubles for other users.

**Comment by Sil3ntVip3r** [comment]
I reproduced the current DNSSEC query-log state and prepared the dependency-side fix in [dnsproxy #519](https://github.com/AdguardTeam/dnsproxy/pull/519).

The original SERVFAIL false-positive path is no longer present: a locally generated/upstream-failure SERVFAIL reports no authenticated data.  A related current false-negative remains, however.  When DNSSEC is enabled and a validating upstream returns AD=true for a valid answer, dnsproxy must clear AD from the wire response if the client did not send AD or DO.  AdGuard Home currently records pctx.Res.AuthenticatedData after that client-specific filtering, so the query log loses a valid upstream DNSSEC result for ordinary clients.

PR #519 exposes the pre-filter response AD state through DNSContext.ResponseAD for fresh upstreams and cache hits, resets it on context reuse, and preserves it across pending-request coalescing.  The eventual AdGuard Home follow-up is then a one-line metadata change in processUpstream:

    dctx.responseAD = pctx.ResponseAD()

Validation completed:

- valid upstream AD with no client AD/DO: wire AD remains false, metadata AD is true;
- cached response keeps the same metadata with one upstream exchange;
- generated SERVFAIL and reused contexts remain false;
- pending-request leader and waiters agree;
- focused regressions pass 20 times under race;
- full dnsproxy proxy package, lint/vulnerability scan, and all supported-OS vet checks pass;
- full AdGuard Home internal/dnsforward package passes under race with the local dependency and one-line consumer change;
- independent adversarial review is clean after two edge-case corrections.

This keeps protocol behavior unchanged while making the Query Log indicator reflect the resolver validation result rather than the client-specific wire filtering.  I will keep the AdGuard Home consumer change isolated until the dependency API can be consumed normally.
